lib/dialogs/settings.jsx

```diff
@@ -38,29 +38,25 @@ export const SettingsDialog = React.createClass({
     });
   },
 
-  onSignOutRequested() {
+  async onSignOutRequested() {
     // Safety first! Check for any unsynced notes before signing out.
     const { onSignOut, noteBucket } = this.props;
     const { notes } = this.props.appState;
     const { getVersion } = noteBucket;
 
-    noteBucket.hasLocalChanges((error, hasChanges) => {
-      if (hasChanges) {
+    if (await noteBucket.hasLocalChanges()) {
+      this.showUnsyncedWarning();
+      return;
+    }
+
+    for (const note of notes) {
+      if ((await getVersion(note)) === 0) {
         this.showUnsyncedWarning();
         return;
       }
+    }
```

Comment:
is this going to change the semantics? will this spread out each async action and then stop getting the versions of others once one is found?

Reply:
Yes, it will do an early abort if it enters the true branch here.

Reply:
Right, this will only do a single async action at a time so presumably it good be less performant in that it won't make the requests in parallel.

Reply:
Another rewrite that preserves the potential parallelization (but throwing an Error is kind of gross to me):

```js
try {
  await Promise.all(notes.map( async note => {
    if ( await getVersion( note ) === 0 ) {
       throw new Error( 'not synced' );
    }
  } ) );
} catch ( error ) {
    this.showUnsyncedWarning();
    return;
}
onSignOut();
```

Reply:
```js
const noteVersions = await Promise.all( notes.map( async note => await getVersion( note ) ) )
const needsSync = noteVersions.some( v => v === 0 )

return needsSync
	? this.showUnsyncedWarning()
	: onSignOut()
```

???

Reply:
The beauty of the `reject`in the original implementation is that it _will_ short circuit any pending async operations.

Reply:
You can get rid of the `async/await` in the closure since `getVersion` returns a `Promise` and it becomes:
```js
const noteVersions = await Promise.all( notes.map( getVersion ) )
```

Reply:
so we could still reject here too, no? I was overlooking that the first `reject()` rejects and abort the entire `Promise.all()`

```js
const noteHasSynced = note => /* identical to before */

try {
  await Promise.all( notes.map( noteHasSynced ) )
  onSignOut()
} catch ( e ) {
  this.showUnsynchedWarning()
}
```

Reply:
If JavaScript catch statements had pattern matching we could check for the expected error type.

The `reject` in `noteHasSynced` should be providing an `Error` instance to `reject`.